File: 2300.咒语和药水的成功对数.cpp
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

/*
 * @lc app=leetcode.cn id=2300 lang=cpp
 *
 * [2300] 咒语和药水的成功对数
 */

 // @lc code=start
class Solution {
    public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success)
    {
        vector<int> res(spells.size());
        vector<int> idx(spells.size());
        iota(idx.begin(), idx.end(), 0);
        sort(idx.begin(), idx.end(), [&] (int a, int b) {
            return spells[a] < spells[b];
            }
        );
        sort(potions.begin(), potions.end(), [] (int a, int b) {
            return a > b;
            }
        );

        for (int i = 0,j=0; i < spells.size(); i++)
        {
            int p = idx[i];
            int v = spells[p];
            while (j<potions.size()&&(long long)potions[j]*v>=success)
            {
                ++j;
            }
            res[p] = j;
        }
        return res;
    }
};
// @lc code=end

File: 2300.咒语和药水的成功对数.go
package leetcode

import "sort"

/*
 * @lc app=leetcode.cn id=2300 lang=golang
 *
 * [2300] 咒语和药水的成功对数
 */

// @lc code=start
func successfulPairs(spells []int, potions []int, success int64) (result []int) {
	sort.Ints(potions)
	n := len(potions)
	for _, spell := range spells {
		index := int(success) / spell
		if int(success)%spell != 0 {
			index++
		}
		result = append(result, n-sort.SearchInts(potions, index))
	}
	return result
}

// @lc code=end
